service/fx.go

```diff
@@ -86,19 +115,29 @@ func NewPersistenceRateLimitingParams(
 func PersistenceMaxQpsFn(
 	maxQps dynamicconfig.IntPropertyFn,
 	globalMaxQps dynamicconfig.IntPropertyFn,
+	lazyLoadedServiceResolver PersistenceLazyLoadedServiceResolver,
 ) persistenceClient.PersistenceMaxQps {
 	return func() int {
-		// if globalMaxQps() > 0 {
-		// 	// TODO: We have a bootstrap issue to correctly find memberCount.  Membership relies on
-		// 	// persistence to bootstrap membership ring, so we cannot have persistence rely on membership
-		// 	// as it will cause circular dependency.
-		// 	// ringSize, err := membershipMonitor.GetMemberCount(serviceName)
-		// 	// if err == nil && ringSize > 0 {
-		// 	// 	avgQuota := common.MaxInt(globalMaxQps()/ringSize, 1)
-		// 	// 	return common.MinInt(avgQuota, maxQps())
-		// 	// }
-		// }
-		return maxQps()
+		// TODO: create lazy loaded version of ClusterAwareQuotaCalculator and
+		// ClusterAwareNamespaceSpecificQuotaCalculator
+		instanceLimit := maxQps()
+
+		value := (*lazyLoadedServiceResolver).Load()
+		if value == nil {
+			return instanceLimit
+		}
+
+		memberCount := value.(membership.ServiceResolver).MemberCount()
+		if memberCount <= 0 {
+			return instanceLimit
+		}
+
+		globalRps := globalMaxQps()
+		if globalRps <= 0 {
+			return instanceLimit
+		}
+
+		return util.Min(instanceLimit, globalRps/memberCount)
```

Comment:
This would be inconsistent with the other logic we have where we just use the global one if it's available and ignore the instance limit. I know it's a bit unintuitive, but that's how we do it. Is there a reason for changing that here?

Reply:
I somehow thought that's what other place doing, I probably mis-read the code before. 

The impl is now using clusterAwareQuotaCalculator, and we will have the consistent behavior.